### ISSUE TITLE
[Bug] Validations not working as expected.

### ISSUE THREAD

**Issue by Sharaf-Mansour** [Blazored/FluentValidation#104]
This is a screenshot from your demo at https://blazored.github.io/FluentValidation/
 For Complex Types, Inputs only shows Validation Messages on submit (not by binding)
![image](https://user-images.githubusercontent.com/55330747/143796917-dc518c5d-432c-4b11-8667-59b17a79247d.png)
Expected Behavior should be this
![image](https://user-images.githubusercontent.com/55330747/143797043-e624645f-05df-4890-9f8c-ed0e9f5b7ec7.png)
I have installed your latest Version and have issue myself here 
![image](https://user-images.githubusercontent.com/55330747/143797283-3fc78139-641e-4f9f-a56d-e30010538a1c.png)
When it should say
![image](https://user-images.githubusercontent.com/55330747/143797323-6b453806-9e16-4914-8106-4734f3040c6f.png)
My Repo can be found here https://github.com/Sharaf-Mansour/CVBuilder.

**Comment by kylepope-ge** [comment]
I ran into this same problem recently and (incorrectly) assumed it was due to the dotnet 6 upgrade of our blazor wasm app that I was doing at the same time. It looks like this is a new issue introduced in version 2.0.3 from fix in PR[#91](https://github.com/Blazored/FluentValidation/pull/91). I'm not exactly sure _why_ that fix causes this new problem, but I can also repoduce this strange behavior where nested properties aren't validated onchange correctly.

Downgrading to 2.0.1 fixed the issue for me.

**Comment by Sharaf-Mansour** [comment]
Thanks for the help, but I am pretty sure that this issue needed to be addressed so it get fixes in the upcoming releases.

**Comment by leonibr** [comment]
> I ran into this same problem recently and (incorrectly) assumed it was due to the dotnet 6 upgrade of our blazor wasm app that I was doing at the same time. It looks like this is a new issue introduced in version 2.0.3 from fix in PR[#91](https://github.com/Blazored/FluentValidation/pull/91). I'm not exactly sure _why_ that fix causes this new problem, but I can also repoduce this strange behavior where nested properties aren't validated onchange correctly.
> 
> Downgrading to 2.0.1 fixed the issue for me.

Thanks for sharing

**Comment by dubluxas** [comment]
Same problem. Does not show messages in child components.

**Comment by cbordeman** [comment]
This solution only partially works.  With components 2 or more levels deep,  _adding_ validation messages works with 2.0.1, but not removing them.

**Comment by Sharaf-Mansour** [comment]
@cbordeman How do you go levels deep ?
## Option 1
```csharp 
RuleFor(x => x.Country).SetValidator(new CountryValidator());
```
## Option 2 
```csharp
RuleFor(x => x.Country).Null().When(x=>x.Country.Id is 0).WithMessage("Country required.");
```
## Option 3
```csharp
RuleFor(x => x.Country.Id).SetValidator(new CountryIDValidator());
```
## Option 4
```csharp
 RuleFor(x => x.Country.Id).NotEqual(0).WithMessage("Country required.");
```

**Comment by cbordeman** [comment]
What I mean is 2 Razor components:

App page (validation works ok)
    <BuyerComponent Buyer="@buyer"> (validation works ok)
        AddressComponent Address="@address"> (only removes errors, doesn't add them)

The BuyerComponent and AddressComponent exist for code reuse and have local properties (Parameters) the razor binds against called Buyer and Address.

**Comment by Sharaf-Mansour** [comment]
Try Splitting it into 2 rules and not chain them.
```csharp
 RuleFor(p => p.Street).MaximumLength(70).WithMessage("Too long");
 RuleFor(p => p.Street).NotEmpty().WithMessage("Required");
```

**Comment by cbordeman** [comment]
I have other fields that are very simple and don't work:

RuleFor(p => p.FirstName).MaximumLength(25).WithMessage("Too long");

I removed the Cascade and broke down the others and it doesn't change anything.

**Comment by Sharaf-Mansour** [comment]
@cbordeman are you sending the parent model or the child model to your EditForm component ? 
Are you using static model or  dependency injection?

**Comment by leonibr** [comment]
Hey, @cbordeman, there is a running sample here: https://blazorrepl.telerik.com/QmaQbQle59JWatqq11 and it is working with 2 levels deep: 
```cs
Person.Address.District 
```
Just hit 'Run' and then 'Save' to see validation working, if you have a different setup maybe you can reproduce it on blazorrepl and share here
![Screenshot - 2022-02-12T120415 754](https://user-images.githubusercontent.com/7917333/153716553-d71f9de8-d62e-4d05-b5a2-232ed615d40f.png)

**Comment by cbordeman** [comment]
@leonibr That's not 2 _blazor_ _components_ deep, that's a single binding in the top level page.  I mean you have 1 page, which uses a component, and that component uses a second component.

**Comment by leonibr** [comment]
@cbordeman , ok then, I refactored to use 2 nested blazor components and it is working properley, maybe you are not passing the top-level `EditContext` down to the components
Check it out: https://blazorrepl.telerik.com/mmEmPcPh34yKeTKU4
![Screenshot - 2022-02-12T143205 947](https://user-images.githubusercontent.com/7917333/153721852-19edb196-edb1-424d-b5ed-9ab793a4a017.png)

**Comment by cbordeman** [comment]
That fiddle is empty.

**Comment by Sharaf-Mansour** [comment]
@cbordeman I have an app that goes 3 or maybe 4 levels deep with validations [here](https://sharaf-mansour.github.io/CVBuilder/)
Source code [here](https://github.com/Sharaf-Mansour/CVBuilder)
It has a custom Address component with custom validation It might be good of use
It is for sure not the simplest sample but it is what I have available and working as expected ATM
Note: Validations are inside ModelValidation Folder

**Comment by cbordeman** [comment]
The problem is that validations remove w/o issue, but don't add (or don't display visually) per-field.  When you tab out of the field (such as removing the text from a .NotEmpty() field.  It only happens on InputTexts on the third level of depth (or page + component + component).  That source doesn't do it, either.

**Comment by leonibr** [comment]
@cbordeman sorry about that: https://blazorrepl.telerik.com/QmOmPQFt37aha3th17

**Comment by Sharaf-Mansour** [comment]
@cbordeman You mean that you want the validation start on input focus not binding ? Or you want it to be removed on blur ?

**Comment by cbordeman** [comment]
Just standard behavior for .NotEmpty().  If there's NOT a validation error showing, then when you remove the value in the input and press TAB, it should appear.  The opposite already works fine (type a value and it disappears).  This works properly on the parent component and directly on the page.

**Comment by Sharaf-Mansour** [comment]
@cbordeman Tabbing from an input does not trigger the validation because it validates on binding or submitting.
If you want it to trigger on Tab or On Focus / On Blur then you will have to create some sort of code that triggers the binding (Shows a single error message) or the Form Submit. (Show all errors messages)

**Comment by Sharaf-Mansour** [comment]
@cbordeman after doing a little research, You can trigger     
```csharp
editContext.Validate();
``` 
Docs can be found [here ](https://docs.microsoft.com/en-us/aspnet/core/blazor/forms-validation?view=aspnetcore-6.0#enable-the-submit-button-based-on-form-validation)

**Comment by cbordeman** [comment]
@Sharaf-Mansour I don't mean an actual <input>, I mean a <InputText> component, so I don't need to manually call .Validate().  My code seems to be identical in structure to the fiddle @leonibr posted.  I don't know why hers doesn't exhibit this behavior!  @leonibr what version of Blazored.FluentValidation are you using?  I'm using 2.0.1 since 2.0.3 introduced a bug.

I added a ValidationSummary to mine and that seems to work as expected, both removing errors as expected.  IT's just the <ValidationMessage below the <InputText..> that doesn't show the message onblur, (unless you manually invoke .Validate() or click the Submit button ofc).

**Comment by cbordeman** [comment]
I've figured it out.  I subclassed both my validator classes.  If I get rid of the subclassing, all works well.

For example, I have AddressValidatorBase and subclasses AddressValidator and AddressValidatorWithRequiredCounty.  Easy enough to fix, but it is a bug I think.  I tried to do this with rulesets earlier, but had no way to change the call with this package.  I also tried passing as a ctor param but FluentValidations couldn't construct it.

**Comment by Sharaf-Mansour** [comment]
I found a bug as well in my own project. 
I made 2 models one inherit the other.
then made 2 validators one for the parent and other for the chilled.
 In the parent I set rule for name to print (Please enter Parent Name)
And for the child name I set (Please enter child name). 
When I submit empty text in child I get please enter child name as expected 
But if I type something and then remove it.... It says Please enter Parent name..
But get fixed on invalid submit to Please enter child name... 
as a easy fix I just changed the msg to just (Please enter name) But I think 
the issue is with the assembly scanning tech not really in the validators.

**Comment by cbordeman** [comment]
Possibly.

**Comment by cbordeman** [comment]
In both cases of my Address and Person validator subclass, I'm setting them explicitly in every instance they're needed, using .SetValidator(new...).  So it's not selecting the wrong classes in my bug, seems like something else.

**Comment by andrewaggb** [comment]
I'm also having this problem.  the issue for me seems to be in this function.  In my case the field is part of a subobject.
So something like
Person.Address.**Street** but I'm using a single validator for the whole thing.  The full model validation works but the ValidateField doesn't work as-is.  If I change the fieldIdentifier.FieldName to be "Address." + fieldIdentifier.FieldName it works as expected.  I think fieldIdentifier has the object name and I know the overall model name so maybe I can do something with that.  I think I'll look at the other two blazor fluent validation projects and see how they implemented ValidateField.

`
private static async void ValidateField(EditContext editContext,
                                                ValidationMessageStore messages,
                                                FieldIdentifier fieldIdentifier,
                                                IServiceProvider serviceProvider,
                                                bool disableAssemblyScanning,
                                                IValidator validator = null)
        {
            **var properties = new[] { fieldIdentifier.FieldName };**
`

**Comment by andrewaggb** [comment]
Well it's probably not the best/most elegant way but this is what I came up with for my particular problem.  Basically if the fieldIdentifier.Model isn't the same object as the editContext.Model I look through properties of the editContext.Model for ones of the same type and see if they are the same object and build a properyname based on that.  I also added a second pass to go one level deeper so I can handle (Model.Class1.Class2.Property).  In my case I wanted it all in a single validator and not a separate validator for class1 and a separate one for class2 as the business rules between it all is quite complex.  I could have reorganized it into separate validators that took the parent(s) or other objects into the separate validators (so I have the required information for the validation rules) and that may have worked without this change.  I also modified this line messages.Add(fieldIdentifier, validationResults.Errors.Where(x => **properties.Contains(x.PropertyName)**).Select(error => error.ErrorMessage)); to check that the PropertyName was one I was expecting.  In my case I had a single validation rule for the root object that was also returned and I didn't want that message displayed on this property.

`private static ConcurrentDictionary<Type, System.Reflection.PropertyInfo[]> TypeProperties = new ConcurrentDictionary<Type, System.Reflection.PropertyInfo[]>();
        private static async void ValidateField(EditContext editContext,
                                                ValidationMessageStore messages,
                                                FieldIdentifier fieldIdentifier,
                                                IServiceProvider serviceProvider,
                                                bool disableAssemblyScanning,
                                                IValidator validator = null)
        {
            string[] properties = new[] { fieldIdentifier.FieldName };
            if (fieldIdentifier.Model != editContext.Model)
            {
                var editContextType = editContext.Model.GetType();
                if (!TypeProperties.TryGetValue(editContextType, out var editContextProperties))
                {
                    editContextProperties = editContextType.GetProperties();
                    TypeProperties.TryAdd(editContextType, editContextProperties);
                }
                var fieldType = fieldIdentifier.Model.GetType();
                bool found = false;
                foreach (var editContextProperty in editContextProperties.Where(x => x.PropertyType == fieldType))
                {
                    try
                    {
                        if (editContextProperty.GetValue(editContext.Model) == fieldIdentifier.Model)
                        {
                            properties = new[] { $"{editContextProperty.Name}.{fieldIdentifier.FieldName}" };
                            found = true;
                            break;
                        }
                    }
                    catch (System.Reflection.TargetException) { }
                    catch (Exception ex)
                    {
                        Console.WriteLine($"{editContextProperty.Name} {ex}");
                    }
                }
                if (!found)
                {
                    foreach (var editContextProperty in editContextProperties.Where(x => x.PropertyType.IsClass))
                    {

                        if (!TypeProperties.TryGetValue(editContextProperty.PropertyType, out var subClass1Properties))
                        {
                            subClass1Properties = editContextProperty.PropertyType.GetProperties();
                            TypeProperties.TryAdd(editContextProperty.PropertyType, subClass1Properties);
                        }
                        foreach (var subClass1Property in subClass1Properties.Where(x => x.PropertyType == fieldType))
                        {
                            try
                            {
                                if (subClass1Property.GetValue(editContextProperty.GetValue(editContext.Model)) == fieldIdentifier.Model)
                                {
                                    properties = new[] { $"{editContextProperty.Name}.{subClass1Property.Name}.{fieldIdentifier.FieldName}" };
                                    found = true;
                                    break;
                                }
                            }
                            catch (System.Reflection.TargetException) { }
                            catch (Exception ex)
                            {
                                Console.WriteLine($"{editContextProperty.Name}.{subClass1Property.Name} {ex}");
                            }
                        }
                        if (found) break;
                    }
                }
            }
            var context = new ValidationContext<object>(editContext.Model, new PropertyChain(), new MemberNameValidatorSelector(properties));

            validator ??= GetValidatorForModel(serviceProvider, editContext.Model, disableAssemblyScanning);

            if (validator is object)
            {
                var validationResults = await validator.ValidateAsync(context);

                messages.Clear(fieldIdentifier);
                messages.Add(fieldIdentifier, validationResults.Errors.Where(x => properties.Contains(x.PropertyName)).Select(error => error.ErrorMessage));

                editContext.NotifyValidationStateChanged();
            }
        }`

**Comment by andrewaggb** [comment]
I ran into an unexpected issue with my code above.  the model passed by the fieldIdentifier wasn't always the one from my edit content.   for fieldIdentifiers I created in my razor components and manually trigger it looks fine but for others it isn't.  If I use the ToFieldIdentifier code , eg var fieldIdentifier = ToFieldIdentifier(editContext, properties[0]); then adding the messages works but then I can't use the object matching technique I was using before (assuming that fieldIdentifier.Model would be part of the editContext) to build the property name.  I'm not sure if I'll get around to looking at it further.
